server/src/main/java/io/crate/metadata/doc/DocTableInfo.java

```diff
@@ -431,10 +433,14 @@ public DynamicReference getDynamic(ColumnIdent ident, boolean forWrite) {
 
         switch (parentPolicy) {
             case DYNAMIC:
-                if (!forWrite) return null;
+                if (!forWrite) {
+                    if (!errorOnUnknownObjectKey) {
+                        throw new UnknownObjectKeyExceptionalControlFlow(ident.sqlFqn());
+                    }
+                    return null;
+                }
                 break;
             case STRICT:
```

Comment:
I'm not sure if these changes are right. I think we still want to allow creating new columns as part of an insert/update, independent of the session setting. Otherwise this would be a breaking change.

Reply:
for insert/update queries, they will break out of the switch and return a `DynamicReference`.  `UnknownObjectKeyExceptionalControlFlow` is dedicated to read operations that also have the session setting set to false.